numba/cuda/cudadrv/devicearray.py

```diff
@@ -183,7 +184,10 @@ def _numba_type_(self):
             layout = 'A'
 
         dtype = numpy_support.from_dtype(self.dtype)
-        return types.Array(dtype, self.ndim, layout)
+        if isinstance(dtype, types.NestedArray):
```

Comment:
I think this change is a leftover from an earlier iteration of the PR - it seems never to take the `if` branch when running the tests.

BTW - I was going to ask if you could explain why the change was needed, or alternatively add some comments I'd prepared below, but it seems like just removing this change is probably the better option. Suggested changes I'd prepared, for reference:

```diff
commit 18907ae4a20237b9c60a20322684767781829f97 (HEAD -> testhound/set_structured_type_dev_array_6114)
Author: Graham Markall <gmarkall@nvidia.com>
Date:   Fri Aug 6 12:49:48 2021 +0100

    [NFC] Add some notes on dtype of nested arrays

diff --git a/numba/cuda/cudadrv/devicearray.py b/numba/cuda/cudadrv/devicearray.py
index d2720022f..80bd05e7c 100644
--- a/numba/cuda/cudadrv/devicearray.py
+++ b/numba/cuda/cudadrv/devicearray.py
@@ -161,33 +161,59 @@ class DeviceNDArrayBase(_devicearray.DeviceArray):
         Magic attribute expected by Numba to get the numba type that
         represents this object.
         """
-        # Typing considerations:
-        #
-        # 1. The preference is to use 'C' or 'F' layout since this enables
-        # hardcoding stride values into compiled kernels, which is more
-        # efficient than storing a passed-in value in a register.
+        dtype = numpy_support.from_dtype(self.dtype)
+
+        # NOTE: the nested array typing is somewhat confusing - at this point
+        # in the code, it feels as though there is one level of indirection too
+        # many in the data structure (if the array's type is a nested array, it
+        # seems to be structured as if it's an array of nested arrays).
         #
-        # 2. If an array is both C- and F-contiguous, prefer 'C' layout as it's
-        # the more likely / common case.
+        # The following two cases differentiate between the scalar dtype and
+        # nested array dtype cases.
         #
-        # 3. If an array is broadcast then it must be typed as 'A' - using 'C'
-        # or 'F' does not apply for broadcast arrays, because the strides, some
-        # of which will be 0, will not match those hardcoded in for 'C' or 'F'
-        # layouts.
-
-        broadcast = 0 in self.strides
-        if self.flags['C_CONTIGUOUS'] and not broadcast:
+        # * For nested arrays, the comments explain the (questionable?)
+        #   rationale behind each assignment.
+        # * For the scalar dtype case, the layout considerations are more
+        #   reasonable, and are commented / documented to give the
+        #   considerations behind the layout computation.
+        if isinstance(dtype, types.NestedArray):
+            # Nested arrays are found within structured types, and are always
+            # C-contiguous.
             layout = 'C'
-        elif self.flags['F_CONTIGUOUS'] and not broadcast:
-            layout = 'F'
+            # Nested arrays have their own dtype member, giving the type of
+            # elements of the nested array. We need the element type (not the
+            # nested array type) for the dtype of the returned array type.
+            dtype = dtype.dtype
+            # The number of dimensions of the array type should be derived from
+            # the shape of the nested array, rather than the shape of the array
+            # holding it.
+            ndim = len(dtype.shape)
         else:
-            layout = 'A'
+            # General (non-nested array) layout considerations:
+            #
+            # 1. The preference is to use 'C' or 'F' layout since this enables
+            # hardcoding stride values into compiled kernels, which is more
+            # efficient than storing a passed-in value in a register.
+            #
+            # 2. If an array is both C- and F-contiguous, prefer 'C' layout as
+            # it's the more likely / common case.
+            #
+            # 3. If an array is broadcast then it must be typed as 'A' - using
+            # 'C' or 'F' does not apply for broadcast arrays, because the
+            # strides, some of which will be 0, will not match those hardcoded
+            # in for 'C' or 'F' layouts.
+
+            broadcast = 0 in self.strides
+            if self.flags['C_CONTIGUOUS'] and not broadcast:
+                layout = 'C'
+            elif self.flags['F_CONTIGUOUS'] and not broadcast:
+                layout = 'F'
+            else:
+                layout = 'A'
 
-        dtype = numpy_support.from_dtype(self.dtype)
-        if isinstance(dtype, types.NestedArray):
-            return types.Array(dtype.dtype, len(dtype.shape), "C")
-        else:
-            return types.Array(dtype, self.ndim, layout)
+            ndim = self.ndim
+
+        return types.Array(dtype, ndim, layout)
 
     @property
     def device_ctypes_pointer(self):
```